src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,286 @@
+        {
+            if (!fl.second[0]->as<ASTFunction>() || fl.second[0]->as<ASTFunction>()->name != inter_func_name)
+                ans.second.push_back(fl.second[0]);
+
+            if (fl.second.size() == 2 && (!fl.second[1]->as<ASTFunction>() || fl.second[1]->as<ASTFunction>()->name != inter_func_name))
+                ans.second.push_back(fl.second[1]);
+        }
+        else
+            throw Exception("did not expect that", ErrorCodes::UNEXPECTED_AST_STRUCTURE);
+        return ans;
+    }
+}
+
+/// rebuilds tree, all scalar values now outside the main func
+void buildTree(ASTFunction * old_tree, std::string& func_name, std::string& intro_func, std::pair<ASTs, ASTs> & tree_comp, std::pair<int, std::string> & flag)
+{
+    ASTFunction copy = *old_tree;
+    ASTs cons_val = tree_comp.first;
+    ASTs non_cons = tree_comp.second;
+
+    old_tree->name = intro_func;
+    for (auto & i : cons_val)
+    {
+        old_tree->arguments->children = {};
+        old_tree->arguments->children.push_back(i);
+
+        old_tree->arguments->children.push_back(makeASTFunction(intro_func));
+        old_tree = old_tree->arguments->children[1]->as<ASTFunction>();
+    }
+
+    if (flag.first == -1)
+        old_tree->name = flag.second;
+    else
+        old_tree->name = func_name;
+
+    if (non_cons.empty())
+        throw Exception("Aggregate function " + func_name + " requires single argument",
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    if (non_cons.size() == 1)
+        old_tree->arguments->children.push_back(non_cons[0]);
+    else
+    {
+        size_t i = 0;
+
+        while (i < non_cons.size() - 2)
+        {
+            old_tree->arguments->children = {};
+            old_tree->children.push_back(non_cons[i]);
+
+            old_tree->arguments->children.push_back(makeASTFunction(intro_func));
+            old_tree = old_tree->arguments->children[1]->as<ASTFunction>();
+            ++i;
+        }
+        old_tree->arguments->children = {non_cons[i], non_cons[i + 1]};
+    }
+}
+
+void sumOptimize(ASTFunction * f_n)
+{
+    std::string sum = "sum";
+    std::string mul = "multiply";
+
+    const auto * function_args = f_n->arguments->as<ASTExpressionList>();
+
+    if (!function_args || function_args->children.size() != 1)
+        throw Exception("Wrong number of arguments for function 'sum' (" + toString(function_args->children.size()) + " instead of 1)",
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    auto * inter_node = f_n->arguments->children[0]->as<ASTFunction>();
+    if (inter_node && inter_node->name == mul)
+    {
+        std::pair<ASTs, ASTs> nodes = findAllConsts(f_n, mul);
+
+        if (nodes.first.empty())
+            return;
+
+        std::pair<int, std::string> cur = {1, "have no opposite func"};
+        buildTree(f_n, sum, mul, nodes, cur);
+    }
+}
+
+void minOptimize(ASTFunction * f_n)
+{
+    std::string min = "min";
```

Comment:
It's possible to make all these consts `static constexpr const` to help compiler optimize the code.